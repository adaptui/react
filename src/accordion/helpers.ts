import * as React from "react";

import { AccordionPanelOptions } from "./AccordionPanel";
import { AccordionTriggerOptions } from "./AccordionTrigger";

export type StringOrNull = string | null;

/**
 * When <AccordionPanel> is used without accordionId:
 *
 *  - First render: getAccordionId will return undefined because options.panels
 * doesn't contain the current panel yet (registerPanel wasn't called yet).
 * Thus registerPanel will be called without groupId (accordionId).
 *
 *  - Second render: options.panels already contains the current panel (because
 * registerPanel was called in the previous render). This means that we'll be
 * able to get the related accordionId with the accordion panel index. Basically,
 * we filter out all the accordions and panels that have already matched. In this
 * phase, registerPanel will be called again with the proper groupId (accordionId).
 *
 *  - In the third render, panel.groupId will be already defined, so we just
 * return it. registerPanel is not called.
 */
export function getAccordionId(options: AccordionPanelOptions) {
  const { panels, id, items } = options;
  const panel = panels?.find(p => p.id === id);
  const accordionId = options.accordionId || panel?.groupId;

  if (accordionId || !panel || !panels || !items) {
    return accordionId;
  }

  const panelIndex = getPanelIndex(panels, panel);
  const accordionsWithoutPanel = getAccordionsWithoutPanel(items, panels);

  return accordionsWithoutPanel[panelIndex]?.id || undefined;
}

function getPanelIndex(
  panels: AccordionPanelOptions["panels"],
  panel: typeof panels[number],
) {
  const panelsWithoutAccordionId = panels.filter(p => !p.groupId);

  return panelsWithoutAccordionId.indexOf(panel);
}

function getAccordionsWithoutPanel(
  accordions: AccordionPanelOptions["items"],
  panels: AccordionPanelOptions["panels"],
) {
  const panelsAccordionIds = panels.map(panel => panel.groupId).filter(Boolean);

  return accordions.filter(
    item => panelsAccordionIds.indexOf(item.id || undefined) === -1,
  );
}

export function isPanelVisible(options: AccordionPanelOptions) {
  const { allowMultiple, selectedId, selectedIds } = options;
  const accordionId = getAccordionId(options);

  if (allowMultiple)
    return accordionId ? selectedIds?.includes(accordionId) : false;

  return accordionId ? selectedId === accordionId : false;
}

export function isAccordionSelected(options: AccordionTriggerOptions) {
  const { id, allowMultiple, selectedId, selectedIds } = options;

  if (!id) return;

  if (allowMultiple) return selectedIds?.includes(id);

  return selectedId === id;
}

export function useAccordionPanelId(options: AccordionTriggerOptions) {
  const { panels, id } = options;

  return React.useMemo(
    () => panels?.find(panel => panel.groupId === id)?.id || undefined,
    [panels, id],
  );
}
