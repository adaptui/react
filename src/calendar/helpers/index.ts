/**
 * All credit goes to [React Spectrum](https://github.com/adobe/react-spectrum)
 * for these utils inspiration
 */
import { RangeValue } from "@react-types/shared";
import { useDateFormatter } from "@react-aria/i18n";

import { toUTCString, setDay } from "../../utils";

export function useWeekDays(weekStart: number) {
  const dayFormatter = useDateFormatter({ weekday: "short" });
  const dayFormatterLong = useDateFormatter({ weekday: "long" });

  return [0, 1, 2, 3, 4, 5, 6].map(index => {
    const dateDay = setDay(Date.now(), (index + weekStart) % 7);

    const day = dayFormatter.format(dateDay);
    const dayLong = dayFormatterLong.format(dateDay);
    return { title: dayLong, abbr: day } as const;
  });
}

export function generateDaysInMonthArray(
  month: number,
  monthStartsAt: number,
  weeksInMonth: number,
  year: number,
) {
  return Array(weeksInMonth)
    .fill(1)
    .reduce((weeks: Date[][], _, weekIndex) => {
      const daysInWeek = [0, 1, 2, 3, 4, 5, 6].reduce(
        (days: Date[], dayIndex) => {
          const day = weekIndex * 7 + dayIndex - monthStartsAt + 2;
          const utcDate = toUTCString(new Date(year, month, day));
          const cellDate = new Date(utcDate);

          return [...days, cellDate];
        },
        [],
      );

      return [...weeks, daysInWeek];
    }, []);
}

export function makeRange(start: Date, end: Date): RangeValue<Date> {
  if (end < start) {
    [start, end] = [end, start];
  }

  return { start, end };
}

export * from "./useWeekStart";
