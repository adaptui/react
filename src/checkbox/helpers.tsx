import * as React from "react";
import { warning } from "reakit-warning";
import { createEvent } from "reakit-utils";

import { CheckboxOptions } from "./Checkbox";

export function getChecked(options: CheckboxOptions) {
  const { checked, value, state } = options;
  if (typeof checked !== "undefined") return checked;

  if (typeof value === "undefined") return !!state;

  const stateProp = Array.isArray(state) ? state : [];

  return stateProp.indexOf(value) !== -1;
}

export function fireChange(
  element: HTMLElement,
  onChange?: React.ChangeEventHandler,
) {
  const event = createEvent(element, "change");

  Object.defineProperties(event, {
    type: { value: "change" },
    target: { value: element },
    currentTarget: { value: element },
  });

  onChange?.(event as any);
}

export function useIndeterminateState(
  ref: React.RefObject<HTMLInputElement>,
  options: CheckboxOptions,
) {
  const { state } = options;

  React.useEffect(() => {
    const element = ref.current;

    if (!element) {
      warning(
        state === "indeterminate",
        "Can't set indeterminate state because `ref` wasn't passed to component.",
        "See https://reakit.io/docs/checkbox/#indeterminate-state",
      );
      return;
    }

    if (state === "indeterminate") {
      element.indeterminate = true;
    } else if (element.indeterminate) {
      element.indeterminate = false;
    }
  }, [state, ref]);
}
