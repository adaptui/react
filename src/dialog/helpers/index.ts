export * from "./DialogBackdropContext";
export * from "./useDisableHoverOutside";
export * from "./useDisclosureRef";
export * from "./useEventListenerOutside";
export * from "./useFocusOnBlur";
export * from "./useFocusOnChildUnmount";
export * from "./useFocusOnHide";
export * from "./useFocusOnShow";
export * from "./useFocusTrap";
export * from "./useHideOnClickOutside";
export * from "./useNestedDialogs";
export * from "./usePortalRef";
