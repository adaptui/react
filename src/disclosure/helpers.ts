import * as React from "react";
import { isSelfTarget } from "reakit-utils";
import { useSafeLayoutEffect } from "@chakra-ui/hooks";

function useLastValue<T>(value: T) {
  const lastValue = React.useRef<T | null>(null);

  useSafeLayoutEffect(() => {
    lastValue.current = value;
  }, [value]);

  return lastValue;
}

export type useAnimationProps = {
  visible: boolean;
};

export const useAnimation = (props: useAnimationProps) => {
  const { visible } = props;

  const [animating, setAnimating] = React.useState(false);
  const lastVisible = useLastValue(visible);
  const visibleHasChanged =
    lastVisible.current != null && lastVisible.current !== visible;
  const [state, setState] = React.useState<TransitionState>(null);
  const raf = React.useRef(0);

  if (!animating && visibleHasChanged) {
    // Sets animating to true when when visible is updated
    setAnimating(true);
  }

  React.useEffect(() => {
    // Double RAF is needed so the browser has enough time to paint the
    // default styles before processing the `data-enter` attribute. Otherwise
    // it wouldn't be considered a transition.
    // See https://github.com/reakit/reakit/issues/643
    raf.current = window.requestAnimationFrame(() => {
      raf.current = window.requestAnimationFrame(() => {
        if (visible) {
          setState("enter");
        } else if (animating) {
          setState("leave");
        } else {
          setState(null);
        }
      });
    });

    return () => window.cancelAnimationFrame(raf.current);
  }, [visible, animating]);

  const stopAnimation = React.useCallback(() => setAnimating(false), []);

  const onEnd = React.useCallback(
    (event: React.SyntheticEvent) => {
      if (!isSelfTarget(event)) return;
      if (!animating) return;

      // Ignores number animated
      stopAnimation();
    },
    [animating, stopAnimation],
  );

  return { state, animating, onEnd };
};

export type useAnimationReturnType = ReturnType<typeof useAnimation>;

export type TransitionState = "enter" | "leave" | null;

export function getElementHeight(
  el: React.RefObject<HTMLElement> | { current?: { scrollHeight: number } },
): string | number {
  if (!el?.current) {
    return "auto";
  }

  return el.current.scrollHeight;
}

export function getElementWidth(
  el: React.RefObject<HTMLElement> | { current?: { scrollWidth: number } },
): string | number {
  if (!el?.current) {
    return "auto";
  }

  return el.current.scrollWidth;
}

// https://github.com/mui-org/material-ui/blob/da362266f7c137bf671d7e8c44c84ad5cfc0e9e2/packages/material-ui/src/styles/transitions.js#L89-L98
export function getAutoSizeDuration(size: number | string): number {
  if (!size || typeof size === "string") {
    return 0;
  }

  const constant = size / 36;

  // https://www.wolframalpha.com/input/?i=(4+%2B+15+*+(x+%2F+36+)+**+0.25+%2B+(x+%2F+36)+%2F+5)+*+10
  return Math.round((4 + 15 * constant ** 0.25 + constant / 5) * 10);
}
