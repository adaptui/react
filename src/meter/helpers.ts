/**
 * The candidate optimum point is the midpoint between the minimum value and
 * the maximum value.
 *
 * @see https://html.spec.whatwg.org/multipage/form-elements.html#the-meter-element:attr-meter-high-8:~:text=boundary.-,The%20optimum%20point
 */
export function getDefaultOptimumValue(min: number, max: number) {
  return max < min ? min : min + (max - min) / 2;
}

type CalculateStatusProps = {
  value: number;
  optimum: number;
  min: number;
  max: number;
  low: number;
  high: number;
};

export const calculateStatus = (props: CalculateStatusProps) => {
  const { value, optimum, min, max, low, high } = props;

  // This check always comes first
  if (isInRange(optimum, low, high)) {
    if (isInRange(value, low, high)) return "safe";
    return "caution";
  }

  if (isInRange(optimum, min, low)) {
    if (isInRange(value, min, low)) return "safe";
    if (value > low && value <= high) return "caution";
    return "danger";
  }

  if (isInRange(optimum, high, max)) {
    if (isInRange(value, high, max)) return "safe";
    if (value < high && value >= low) return "caution";
    return "danger";
  }

  // A safe return
  return "safe";
};

export const isInRange = (value: number, min: number, max: number) =>
  value >= min && value <= max;
