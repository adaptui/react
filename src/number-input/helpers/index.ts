import {
  countDecimalPlaces,
  StringOrNumber,
  toPrecision,
} from "@chakra-ui/utils";
import { KeyboardEvent } from "react";

export function parse(value: StringOrNumber) {
  return parseFloat(value.toString().replace(/[^\w.-]+/g, ""));
}

export function getDecimalPlaces(value: number, step: number) {
  return Math.max(countDecimalPlaces(step), countDecimalPlaces(value));
}

export function cast(value: StringOrNumber, step: number, precision?: number) {
  const decimalPlaces = getDecimalPlaces(parse(value), step);
  return toPrecision(parse(value), precision ?? decimalPlaces);
}

const FLOATING_POINT_REGEX = /^[Ee0-9+\-.]$/;

/**
 * Determine if a character is a DOM floating point character
 * @see https://www.w3.org/TR/2012/WD-html-markup-20120329/datatypes.html#common.data.float
 */
export function isFloatingPointNumericCharacter(character: string) {
  return FLOATING_POINT_REGEX.test(character);
}

/**
 * Determine if the event is a valid numeric keyboard event.
 * We use this so we can prevent non-number characters in the input
 */
export function isValidNumericKeyboardEvent(event: React.KeyboardEvent) {
  if (event.key == null) return true;

  const isModifierKey = event.ctrlKey || event.altKey || event.metaKey;

  if (isModifierKey) {
    return true;
  }

  const isSingleCharacterKey = event.key.length === 1;

  if (!isSingleCharacterKey) {
    return true;
  }

  return isFloatingPointNumericCharacter(event.key);
}

export function getStepFactor(event: KeyboardEvent) {
  let ratio = 1;

  if (event.metaKey || event.ctrlKey) {
    ratio = 0.1;
  }

  if (event.shiftKey) {
    ratio = 10;
  }

  return ratio;
}

export * from "./useSpinner";
