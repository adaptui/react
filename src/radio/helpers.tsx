import React from "react";
import { createEvent } from "reakit-utils";

import { RadioOptions } from "./Radio";

export function getChecked(options: RadioOptions) {
  const { checked, value, state } = options;
  if (typeof checked !== "undefined") return checked;

  return typeof value !== "undefined" && state === value;
}

export function useInitialChecked(options: RadioOptions) {
  const [initialChecked] = React.useState(() => getChecked(options));
  const { id, currentId, setCurrentId } = options;
  const [initialCurrentId] = React.useState(currentId);

  React.useEffect(() => {
    if (initialChecked && id && initialCurrentId !== id) {
      setCurrentId?.(id);
    }
  }, [initialChecked, id, setCurrentId, initialCurrentId]);
}

export function fireChange(
  element: HTMLElement,
  onChange?: React.ChangeEventHandler,
) {
  const event = createEvent(element, "change");

  Object.defineProperties(event, {
    type: { value: "change" },
    target: { value: element },
    currentTarget: { value: element },
  });
  onChange?.(event as any);
}
