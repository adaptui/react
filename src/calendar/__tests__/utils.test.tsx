import MockDate from "mockdate";
import { renderHook } from "@testing-library/react-hooks";

import { generateDaysInMonthArray, makeRange, useWeekDays } from "../helpers";

describe("Calendar Utils", () => {
  test("makeRange", () => {
    const range = makeRange(
      new Date(1999, 4, 4, 0, 0),
      new Date(2020, 4, 4, 0, 0),
    );
    expect(range.start.toISOString()).toEqual("1999-05-03T18:30:00.000Z");
    expect(range.end.toISOString()).toEqual("2020-05-03T18:30:00.000Z");
  });

  test("useWeekDays", () => {
    // MIND THE BLOCK SCOPE!
    {
      const {
        result: { current },
      } = renderHook(() => useWeekDays(0));

      expect(current).toStrictEqual([
        { abbr: "Sun", title: "Sunday" },
        { abbr: "Mon", title: "Monday" },
        { abbr: "Tue", title: "Tuesday" },
        { abbr: "Wed", title: "Wednesday" },
        { abbr: "Thu", title: "Thursday" },
        { abbr: "Fri", title: "Friday" },
        { abbr: "Sat", title: "Saturday" },
      ]);
    }
    {
      const {
        result: { current },
      } = renderHook(() => useWeekDays(2));

      expect(current).toStrictEqual([
        { abbr: "Tue", title: "Tuesday" },
        { abbr: "Wed", title: "Wednesday" },
        { abbr: "Thu", title: "Thursday" },
        { abbr: "Fri", title: "Friday" },
        { abbr: "Sat", title: "Saturday" },
        { abbr: "Sun", title: "Sunday" },
        { abbr: "Mon", title: "Monday" },
      ]);
    }
  });

  test("generateDaysInMonthArray", () => {
    MockDate.set(new Date("2020-02-01T11:30:00.000Z"));
    const days = generateDaysInMonthArray(1, 0, 7, 2020);

    expect(days).toMatchInlineSnapshot(`
      Array [
        Array [
          2020-02-01T11:30:00.000Z,
          2020-02-02T11:30:00.000Z,
          2020-02-03T11:30:00.000Z,
          2020-02-04T11:30:00.000Z,
          2020-02-05T11:30:00.000Z,
          2020-02-06T11:30:00.000Z,
          2020-02-07T11:30:00.000Z,
        ],
        Array [
          2020-02-08T11:30:00.000Z,
          2020-02-09T11:30:00.000Z,
          2020-02-10T11:30:00.000Z,
          2020-02-11T11:30:00.000Z,
          2020-02-12T11:30:00.000Z,
          2020-02-13T11:30:00.000Z,
          2020-02-14T11:30:00.000Z,
        ],
        Array [
          2020-02-15T11:30:00.000Z,
          2020-02-16T11:30:00.000Z,
          2020-02-17T11:30:00.000Z,
          2020-02-18T11:30:00.000Z,
          2020-02-19T11:30:00.000Z,
          2020-02-20T11:30:00.000Z,
          2020-02-21T11:30:00.000Z,
        ],
        Array [
          2020-02-22T11:30:00.000Z,
          2020-02-23T11:30:00.000Z,
          2020-02-24T11:30:00.000Z,
          2020-02-25T11:30:00.000Z,
          2020-02-26T11:30:00.000Z,
          2020-02-27T11:30:00.000Z,
          2020-02-28T11:30:00.000Z,
        ],
        Array [
          2020-02-29T11:30:00.000Z,
          2020-03-01T11:30:00.000Z,
          2020-03-02T11:30:00.000Z,
          2020-03-03T11:30:00.000Z,
          2020-03-04T11:30:00.000Z,
          2020-03-05T11:30:00.000Z,
          2020-03-06T11:30:00.000Z,
        ],
        Array [
          2020-03-07T11:30:00.000Z,
          2020-03-08T11:30:00.000Z,
          2020-03-09T11:30:00.000Z,
          2020-03-10T11:30:00.000Z,
          2020-03-11T11:30:00.000Z,
          2020-03-12T11:30:00.000Z,
          2020-03-13T11:30:00.000Z,
        ],
        Array [
          2020-03-14T11:30:00.000Z,
          2020-03-15T11:30:00.000Z,
          2020-03-16T11:30:00.000Z,
          2020-03-17T11:30:00.000Z,
          2020-03-18T11:30:00.000Z,
          2020-03-19T11:30:00.000Z,
          2020-03-20T11:30:00.000Z,
        ],
      ]
    `);

    MockDate.reset();
  });
});
