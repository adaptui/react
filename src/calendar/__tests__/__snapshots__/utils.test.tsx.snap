// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar Utils generateDaysInMonthArray 1`] = `
Array [
  Array [
    2020-02-01T11:30:00.000Z,
    2020-02-02T11:30:00.000Z,
    2020-02-03T11:30:00.000Z,
    2020-02-04T11:30:00.000Z,
    2020-02-05T11:30:00.000Z,
    2020-02-06T11:30:00.000Z,
    2020-02-07T11:30:00.000Z,
  ],
  Array [
    2020-02-08T11:30:00.000Z,
    2020-02-09T11:30:00.000Z,
    2020-02-10T11:30:00.000Z,
    2020-02-11T11:30:00.000Z,
    2020-02-12T11:30:00.000Z,
    2020-02-13T11:30:00.000Z,
    2020-02-14T11:30:00.000Z,
  ],
  Array [
    2020-02-15T11:30:00.000Z,
    2020-02-16T11:30:00.000Z,
    2020-02-17T11:30:00.000Z,
    2020-02-18T11:30:00.000Z,
    2020-02-19T11:30:00.000Z,
    2020-02-20T11:30:00.000Z,
    2020-02-21T11:30:00.000Z,
  ],
  Array [
    2020-02-22T11:30:00.000Z,
    2020-02-23T11:30:00.000Z,
    2020-02-24T11:30:00.000Z,
    2020-02-25T11:30:00.000Z,
    2020-02-26T11:30:00.000Z,
    2020-02-27T11:30:00.000Z,
    2020-02-28T11:30:00.000Z,
  ],
  Array [
    2020-02-29T11:30:00.000Z,
    2020-03-01T11:30:00.000Z,
    2020-03-02T11:30:00.000Z,
    2020-03-03T11:30:00.000Z,
    2020-03-04T11:30:00.000Z,
    2020-03-05T11:30:00.000Z,
    2020-03-06T11:30:00.000Z,
  ],
  Array [
    2020-03-07T11:30:00.000Z,
    2020-03-08T11:30:00.000Z,
    2020-03-09T11:30:00.000Z,
    2020-03-10T11:30:00.000Z,
    2020-03-11T11:30:00.000Z,
    2020-03-12T11:30:00.000Z,
    2020-03-13T11:30:00.000Z,
  ],
  Array [
    2020-03-14T11:30:00.000Z,
    2020-03-15T11:30:00.000Z,
    2020-03-16T11:30:00.000Z,
    2020-03-17T11:30:00.000Z,
    2020-03-18T11:30:00.000Z,
    2020-03-19T11:30:00.000Z,
    2020-03-20T11:30:00.000Z,
  ],
]
`;

exports[`Calendar Utils useWeekDays 1`] = `
Array [
  Object {
    "abbr": "Sun",
    "title": "Sunday",
  },
  Object {
    "abbr": "Mon",
    "title": "Monday",
  },
  Object {
    "abbr": "Tue",
    "title": "Tuesday",
  },
  Object {
    "abbr": "Wed",
    "title": "Wednesday",
  },
  Object {
    "abbr": "Thu",
    "title": "Thursday",
  },
  Object {
    "abbr": "Fri",
    "title": "Friday",
  },
  Object {
    "abbr": "Sat",
    "title": "Saturday",
  },
]
`;

exports[`Calendar Utils useWeekDays 2`] = `
Array [
  Object {
    "abbr": "Tue",
    "title": "Tuesday",
  },
  Object {
    "abbr": "Wed",
    "title": "Wednesday",
  },
  Object {
    "abbr": "Thu",
    "title": "Thursday",
  },
  Object {
    "abbr": "Fri",
    "title": "Friday",
  },
  Object {
    "abbr": "Sat",
    "title": "Saturday",
  },
  Object {
    "abbr": "Sun",
    "title": "Sunday",
  },
  Object {
    "abbr": "Mon",
    "title": "Monday",
  },
]
`;
